src/theory/strings/core_solver.cpp

```diff
@@ -952,7 +952,7 @@ void CoreSolver::processNEqc(std::vector<NormalForm>& normal_forms)
         {
           return;
         }
-        else if (!pinfer.empty() && pinfer.back().d_id == 1)
+        else if (!pinfer.empty() && pinfer.back().d_id == Inference::INFER_EMP)
```

Comment:
There is an issue here, this is actually an optimization that says effectively "If the inference has a minimal id, then we cannot find a better inference, so break". This changes the sematics of this optimization. Maybe we can just delete this or otherwise fix this in a clear way.

Reply:
As discussed offline, I've investigated the impact of this and it seems negligible. I have submitted #4129 to remove this heuristic.